File: backend/product-service/src/main/java/com/ecommerce/product/service/ProductService.java
package com.ecommerce.product.service;

import com.ecommerce.product.dto.request.CreateProductRequest;
import com.ecommerce.product.dto.request.UpdateProductRequest;
import com.ecommerce.product.dto.response.ProductResponse;

import java.math.BigDecimal;
import java.util.List;

public interface ProductService {

    ProductResponse createProduct(CreateProductRequest request);

    ProductResponse updateProduct(Long id, UpdateProductRequest request);

    void deleteProduct(Long id);

    ProductResponse getProductById(Long id);

    List<ProductResponse> getAllProducts();

    List<ProductResponse> getProductsByCategory(Long categoryId);

    List<ProductResponse> searchProducts(String keyword);

    List<ProductResponse> getProductsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice);

    List<ProductResponse> getProductsByCategoryAndPriceRange(
            Long categoryId, BigDecimal minPrice, BigDecimal maxPrice);
}

File: frontend/react-app/src/services/productService.js
import axiosInstance from '../utils/axiosConfig'

const productService = {

  getAllProducts: async () => {
    const response = await axiosInstance.get('/api/v1/products')
    return response.data
  },

  getProductById: async (id) => {
    const response = await axiosInstance.get(`/api/v1/products/${id}`)
    return response.data
  },

  searchProducts: async (keyword) => {
    const response = await axiosInstance.get(`/api/v1/products/search?keyword=${keyword}`)
    return response.data
  },

  getProductsByCategory: async (categoryId) => {
    const response = await axiosInstance.get(`/api/v1/products/category/${categoryId}`)
    return response.data
  },

  filterProducts: async (categoryId, minPrice, maxPrice) => {
    let url = '/api/v1/products/filter?'
    if (categoryId) url += `categoryId=${categoryId}&`
    if (minPrice) url += `minPrice=${minPrice}&`
    if (maxPrice) url += `maxPrice=${maxPrice}`
    const response = await axiosInstance.get(url)
    return response.data
  },

  getAllCategories: async () => {
    const response = await axiosInstance.get('/api/v1/categories')
    return response.data
  },
}

export default productService